src/communication/v2/listener.hpp

```diff
@@ -0,0 +1,141 @@
+// licenses/APL.txt.
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <memory>
+#include <string_view>
+#include <thread>
+#include <unordered_set>
+
+#include <spdlog/spdlog.h>
+#include <boost/asio/io_context.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/system/detail/error_code.hpp>
+
+#include "communication/context.hpp"
+#include "communication/v2/pool.hpp"
+#include "communication/v2/session.hpp"
+#include "utils/spin_lock.hpp"
+#include "utils/synchronized.hpp"
+
+namespace memgraph::communication::v2 {
+
+template <class TSession, class TSessionData>
+class Listener final : public std::enable_shared_from_this<Listener<TSession, TSessionData>> {
+  using tcp = boost::asio::ip::tcp;
+  using SessionHandler = Session<TSession, TSessionData>;
+
+ public:
+  Listener(const Listener &) = delete;
+  Listener(Listener &&) = delete;
+  Listener &operator=(const Listener &) = delete;
+  Listener &operator=(Listener &&) = delete;
+  ~Listener() {}
+
+  template <typename... Args>
+  static std::shared_ptr<Listener> Create(Args &&...args) {
+    return std::shared_ptr<Listener>{new Listener(std::forward<Args>(args)...)};
+  }
+
+  void Start() { DoAccept(); }
+
+  bool IsRunning() const noexcept { return alive_.load(std::memory_order_relaxed); }
+
+ private:
+  Listener(IOContextThreadPool &io_thread_context_pool, TSessionData *data, ServerContext *server_context,
+           tcp::endpoint &endpoint, const std::string_view service_name, const int inactivity_timeout_sec)
+      : io_thread_context_pool_(io_thread_context_pool),
+        data_(data),
+        server_context_(server_context),
+        acceptor_(io_thread_context_pool.GetIOContext()),
+        endpoint_{endpoint},
+        service_name_{service_name},
+        inactivity_timeout_sec_{inactivity_timeout_sec} {
+    boost::system::error_code ec;
+    // Open the acceptor
+    acceptor_.open(endpoint.protocol(), ec);
+    if (ec) {
+      OnError(ec, "open");
+      return;
+    }
+
+    // Allow address reuse
+    acceptor_.set_option(boost::asio::socket_base::reuse_address(true), ec);
+    if (ec) {
+      OnError(ec, "set_option");
+      return;
+    }
+
+    // Bind to the server address
+    acceptor_.bind(endpoint, ec);
+    if (ec) {
+      spdlog::error(
+          utils::MessageWithLink("Cannot bind to socket on endpoint {}.", endpoint, "https://memgr.ph/socket"));
+      OnError(ec, "bind");
+      return;
+    }
+
+    acceptor_.listen(boost::asio::socket_base::max_listen_connections, ec);
+    if (ec) {
+      OnError(ec, "listen");
+      return;
+    }
+  }
+
+  void DoAccept() {
+    acceptor_.async_accept(io_thread_context_pool_.GetIOContext(),
+                           [shared_this = this->shared_from_this()](auto ec, auto &&socket) {
+                             shared_this->OnAccept(ec, std::forward<decltype(socket)>(socket));
```

Comment:
Some explanations here:

The signature of `OnAccept` is: `OnAccept(boost::system::error_code ec, tcp::socket socket)`

If socket is an `rvalue` reference, then `std::forward` casts it to an rvalue and then moved to the `socket` parameter of `OnAccept`. 

If the socket is an `lvalue` reference, then std::forward does nothing, the result is an lvalue reference and the `socket` gets copied to the socket parameter of `OnAccept`. 

Therefore, with the above-suggested changes, if you now accept by value, then if the result is an `rvalue` reference, the move constructor will be called, and then the socket will be moved to `OnAccept`. That's two moves. 

If the the passed value is an lvalue, then socket will be initialized with a copy which will then be moved to `OnAccept`. 

Therefore, I think `std::forward` is not necessarry for this case.

Reply:
The reason I would keep `std::forward` is that we are dealing with forwarding reference (socket&&), and I would always expect from `async_accept` to make the right call do we need to copy the socket or do we just reuse the given one since if we do `std::move` we can steal the ownership unintentionally, like its shown in this example (https://clang.llvm.org/extra/clang-tidy/checks/bugprone-move-forwarding-reference.html) (clang-tidy told me that)